### PR TITLE
Add retry policy to ProfileOptimization test

### PR DESCRIPTION
**Pull request by MarcoRossignoli**
fixes https://github.com/dotnet/corefx/issues/31792

~~Added retry policy. We retry n times for max \~20 seconds before report test fail.~~

/cc @danmosemsft @brianrob @Sunny-pu

### CODE REVIEW

**Comment by MarcoRossignoli** [comment]
> Assumption is that it just hasn't had a chance to start and write the file in the milliseconds it currently had before we look for it

Understood...my idea was different...i thought that StartProfile is async and sometimes process shutdown before file is written(and won't be written anymore), so i retry with some delay. However if you think that is not possibile i retry only asserts...do you agree?

**Comment by brianrob** [comment]
I don't think that you should need any retry logic around checking if the file exists as long as you to do the following:
 + Run the application.
 + Wait for the application to exit.
 + Check for existence of the file.

```ProfileOptimization.StartProfile``` is synchronous for setup.  The async behavior is starts when background JIT occurs.

**Comment by MarcoRossignoli** [comment]
@brianrob this is original test https://github.com/dotnet/corefx/blob/23fab4631ed927f5657e4ab4a6ce4d45d08bcf7a/src/System.Runtime.Extensions/tests/System/Runtime/ProfileOptimization.netcoreapp.cs
Run the application. -> Remote invoker
Wait for the application to exit. -> Dispose 
Check for existence of the file. -> Asserts
Where do you think is the issue?What i'm missing?

**Comment by brianrob** [comment]
@MarcoRossignoli I don't see anything obviously wrong with this test.  Are you able to reproduce the failure?  Based on what I see, it just started failing, and it is failing deterministically but just for RedHat.  I'm wondering if the file path being used can't be written.  MCJ would just silently fail if this is the case (by design).

**Comment by MarcoRossignoli** [comment]
> MCJ would just silently fail if this is the case (by design).

Sorry...what does MCJ mean?

**Comment by brianrob** [comment]
MCJ == Multi-Core JIT.  It's one of the names for the ProfileOptimization feature.

**Comment by MarcoRossignoli** [comment]
> MCJ == Multi-Core JIT. It's one of the names for the ProfileOptimization feature.

Thank's for clarification.

> Are you able to reproduce the failure?

No...i don't have a Red Hat vm only Ubuntu, i'll try to install and test on Red Hat distro. Or i could add some asserts to test permission

File: src/System.Runtime.Extensions/tests/System/Runtime/ProfileOptimization.netcoreapp.cs [hunk 1]
```diff
@@ -17,7 +17,15 @@ public void ProfileOptimization_CheckFileExists()
...
             RemoteInvoke(profileFilePath =>
             {
-                ProfileOptimization.SetProfileRoot(Path.GetDirectoryName(profileFilePath));
+                string directoryName = Path.GetDirectoryName(profileFilePath);
+
+                // after test fail tracked by https://github.com/dotnet/corefx/issues/31792
+                // we suspect that the reason is something related to write permission to the location
+                // to prove that we added a simple write to file in same location of profile file directory path
+                // ProfileOptimization/Multi-Core JIT could fail silently
+                File.WriteAllText(Path.Combine(directoryName, Path.GetRandomFileName()), "42");
```
Comment (MarcoRossignoli) [inline src/System.Runtime.Extensions/tests/System/Runtime/ProfileOptimization.netcoreapp.cs]: @brianrob added write test with comment before start profiling.

Comment (brianrob) [reply src/System.Runtime.Extensions/tests/System/Runtime/ProfileOptimization.netcoreapp.cs]: @MarcoRossignoli thanks for adding this.  Can you also write the full path to the test file to the console?  That way, if it does fail, we will know exactly where it tried to write to.

File: src/System.Runtime.Extensions/tests/System/Runtime/ProfileOptimization.netcoreapp.cs [hunk 2]
```diff
@@ -5,22 +5,36 @@
...
+
+        public ProfileOptimizationTest(ITestOutputHelper output) => _output = output;
+
         [Fact]
         public void ProfileOptimization_CheckFileExists()
         {
             string tmpProfileFilePath = GetTestFileName();
+            string tmpTestFileName = Path.Combine(Path.GetDirectoryName(tmpProfileFilePath), Path.GetRandomFileName());
+
+            _output.WriteLine($"We'll test write permission on path '{tmpTestFileName}'");
```
Comment (MarcoRossignoli) [inline src/System.Runtime.Extensions/tests/System/Runtime/ProfileOptimization.netcoreapp.cs]: @brianrob @danmosemsft  added some console logs using xunit helper...it doesn't work on my local...is this expected?If so do i use `Console.WriteLine`?

Comment (MarcoRossignoli) [reply src/System.Runtime.Extensions/tests/System/Runtime/ProfileOptimization.netcoreapp.cs]: ping @Anipik @brianrob @danmosemsft

**Comment by danmosemsft** [comment]
@MarcoRossignoli sorry missed this. We don't use ITestOutputHelper anywhere so it doesn't surprise me particularly that it might not work. If you want to gather data locally/in CI I use Console.WriteLine, but we would not check in tests like that.

Does that help?

@ViktorHofer may know why it doesn't work.

**Comment by ViktorHofer** [comment]
> @ViktorHofer may know why it doesn't work.

What doesn't work?

**Comment by danmosemsft** [comment]
@ViktorHofer apparently ITestOutputHelper isn't working for him.

**Comment by MarcoRossignoli** [comment]
> Does that help?

@danmosemsft i added console write suggested by @brianrob https://github.com/dotnet/corefx/pull/31810#discussion_r211409220 it isn't foundamental for test.
@ViktorHofer if i run 
```
...\corefx\src\System.Runtime.Extensions\tests (potest -> origin)
λ msbuild /t:rebuildandtest /v:m "/p:XunitOptions=-method System.Runtime.Tests.ProfileOptimizationTest.ProfileOptimization_CheckFileExists"
```

i get:
```
Discovering: System.Runtime.Extensions.Tests
  Discovered:  System.Runtime.Extensions.Tests
  Starting:    System.Runtime.Extensions.Tests
  Finished:    System.Runtime.Extensions.Tests

  === TEST EXECUTION SUMMARY ===
     System.Runtime.Extensions.Tests  Total: 1, Errors: 0, Failed: 0, Skipped: 0, Time: 0,311s
  ----- end 11:41:33,01 ----- exit code 0 ----------------------------------------------------------
```

I'm expecting console out from https://github.com/dotnet/corefx/pull/31810/files?utf8=%E2%9C%93&diff=split#diff-fa3e09273bf42f1c5b9a2ff25e6f7a3dR24

**Comment by danmosemsft** [comment]
I think @safern may have told me something about console output during xunit runs....

**Comment by ViktorHofer** [comment]
Afaik that's by design: https://github.com/xunit/xunit/issues/720#issuecomment-409873951

**Comment by danmosemsft** [comment]
Thanks @MarcoRossignoli I"ll make a note to check in a few days.

**Comment by danmosemsft** [comment]
Opened new PR

**Pull request merged**